### DISCUSSION TITLE
OpentelemetryLoggingExtensions with Additional LoggingProviders?

### DISCUSSION THREAD

**Discussion by sunildatla** [open-telemetry/opentelemetry-dotnet#5207] [Q&A]
Hi  , 
I see that OpentelemetryLoggingExtensions now have UseTelemetry methods (internal still) which is looks like supporting to take additional loggingproviders other than opentelemetryloggingprovider 

1. This looks to be behind an experimental flag . Is there a timeline when this can be available ?
2. Is there any pre-release package i can use to test this experimental methods
3. I am really interested to inject my own LoggingProviders to send the log to my custom loggingprovider before it goes to opentelemetryloggingprovider and processors . Will this be supported if this feature becomes GA?

tagging @CodeBlanch as you checked in this change .

**Comment by sunildatla** [comment]
@CodeBlanch  re-tagging in case you missed the question

**Comment by CodeBlanch** [comment]
@sunildatla 

> I am really interested to inject my own LoggingProviders to send the log to my custom loggingprovider before it goes to opentelemetryloggingprovider and processors . Will this be supported if this feature becomes GA?

I'm not totally following what you want to do, sorry. Can you give me a little more detail?

* If you want to just send off logs to some other backend...
 
   You can do that today with .NET's API...

   ```csharp
        using var loggerFactory = LoggerFactory.Create(builder => builder
            .AddProvider(new MyCustomLoggerProvider()) // Implement ILoggerProvider
            .AddOpenTelemetry());
   ```

   ...or OTel API...

   ```csharp
        using var loggerFactory = LoggerFactory.Create(builder => builder
            .AddOpenTelemetry(logging => logging
                .AddProcessor(new BatchLogRecordExportProcessor(new MyCustomLogRecordExporter())) // Implement BaseExporter<LogRecord>
                .AddOtlpExporter()));
   ```

* If you want to mess with the `LogRecord`s before they are shipped off, add a processor:

   ```csharp
        using var loggerFactory = LoggerFactory.Create(builder => builder
            .AddOpenTelemetry(logging => logging
                .AddProcessor(new MyCustomLogRecordProcessor()) // Implement BaseProcessor<LogRecord>
                .AddOtlpExporter()));
   ```

The new APIs don't really change these use cases at all.

**Reply by sunildatla** [reply]
@CodeBlanch  , thanks for the response , is adding the processor only way to mess with LogRecord? can we add some sort of Provider in between?

We wanted to modify the schema of the LogRecord and currently doing that with processor , i am worried about the perf with processor usage.

The reason i am asking is i see Microsoft.Extensions.telemetry package is doing something without processor using ExtendedLogger for enriching the Logs(but this could be some static values from startup)
 , i am wondering if something similar can be done. Any recommendations?

https://github.com/dotnet/extensions/blob/f75874b0ffa13c485598a0bdb5832ba9ee54cb03/src/Libraries/Microsoft.Extensions.Telemetry/Logging/ExtendedLogger.cs

**Reply by CodeBlanch** [reply]
@sunildatla 

If you want to add some static values for the process which will be applied to all `LogRecord`s the OTel way to do that is by using `Resource` API:

```csharp
        using var loggerFactory = LoggerFactory.Create(builder => builder
            .AddOpenTelemetry(logging => logging
                .SetResourceBuilder(ResourceBuilder.CreateDefault()
                    .AddAttributes(new Dictionary<string, object>()
                    {
                        ["StaticAttribute1"] = "value1",
                        ["StaticAttribute2"] = "value2",
                    }))
                .AddOtlpExporter()));
```

If you need to do something selectively or using some context the OTel way to do that is via processor. There is an example of doing a [processor for redaction](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/logs/redaction), you could build off of that for enrichment. I'll try to spin up a dedicated doc/example for doing enrichment when I have a cycle or two.

**Reply by CodeBlanch** [reply]
@sunildatla 

> I'll try to spin up a dedicated doc/example for doing enrichment when I have a cycle or two.

Actually, I don't think I'm going to do that yet. There is a good enrichment story for `ILogger` which is _scopes_. Those should work for most users without needing to roll a custom processor. Here is how that looks:

```csharp
using var loggerFactory = LoggerFactory.Create(builder =>
{
    builder.AddOpenTelemetry(options =>
    {
        options.IncludeScopes = true;

        options.AddConsoleExporter();
    });
});

using var logger = loggerFactory.CreateLogger<Program>();

using var scope = logger.BeginScope(
    new Dictionary<string, object>()
    {
        ["enrichedKey1"] = "value1",
    });

logger.FoodPriceChanged("Apple", 9.99);
```

When you emit a log any scopes open will be present on the `LogRecord` if `IncludeScopes` is `true`. It is up to the exporter to handle these correctly so depending on the exporter you are using YMMV.
